src/HtmlProcessor/CssVariableEvaluator.php

```diff
@@ -0,0 +1,197 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Pelago\Emogrifier\HtmlProcessor;
+
+use Pelago\Emogrifier\Utilities\DeclarationBlockParser;
+use Pelago\Emogrifier\Utilities\Preg;
+
+/**
+ * This class can evaluate CSS custom properties that are defined and used in inline style attributes.
+ */
+class CssVariableEvaluator extends AbstractHtmlProcessor
+{
+    /**
+     * temporary collection used by {@see replaceVariablesInDeclarations} and callee methods
+     *
+     * @var array<non-empty-string, string>
+     */
+    private $currentVariableDefinitions = [];
+
+    /**
+     * Replaces all CSS custom property references in inline style attributes with their corresponding values where
+     * defined in inline style attributes (either from the element itself or the nearest ancestor).
+     *
+     * @throws \UnexpectedValueException
+     *
+     * @return $this
+     */
+    public function evaluateVariables(): self
+    {
+        return $this->evaluateVaraiblesInElementAndDescendants($this->getHtmlElement(), []);
+    }
+
+    /**
+     * @param array<non-empty-string, string> $declarations
+     *
+     * @return array<non-empty-string, string>
+     */
+    private function getVaraibleDefinitionsFromDeclarations(array $declarations): array
+    {
+        return \array_filter(
+            $declarations,
+            static function (string $key): bool {
+                return \substr($key, 0, 2) === '--';
+            },
+            ARRAY_FILTER_USE_KEY
+        );
+    }
+
+    /**
+     * Callback function for {@see replaceVariablesInPropertyValue} performing regular expression replacement.
+     *
+     * @param array<int, string> $matches
+     */
+    private function getPropertyValueReplacement(array $matches): string
+    {
+        $variableName = $matches[1];
+        if (isset($this->currentVariableDefinitions[$variableName])) {
+            return $this->currentVariableDefinitions[$variableName];
+        } else {
+            $fallbackValueSeparator = $matches[2] ?? '';
+            if ($fallbackValueSeparator !== '') {
+                $fallbackValue = $matches[3];
+                // The fallback value may use other CSS variables, so recurse
+                return $this->replaceVariablesInPropertyValue($fallbackValue);
+            } else {
```

Comment:
Ditto. (I'm not a big fan of early returns that are not guard clauses, though, and would prefer to make this if/else construct not have early returns.)

Reply:
Ditto.